runner/src/mill/runner/Watching.scala

```diff
@@ -15,72 +15,194 @@ import scala.annotation.tailrec
+      colors: Colors,
+      useNotify: Boolean,
+      serverDir: os.Path
   ): Boolean = {
     setIdle(true)
-    val watchedPaths = watched.collect { case p: Watchable.Path => p.p.path }
-    val watchedValues = watched.size - watchedPaths.size
+    val (watchedPollables, watchedPathsSeq) = watched.partitionMap {
+      case w: Watchable.Value => Left(w)
+      case p: Watchable.Path => Right(p)
+    }
+    val watchedPathsSet = watchedPathsSeq.iterator.map(p => p.p.path).toSet
+    val watchedValueCount = watched.size - watchedPathsSeq.size
 
-    val watchedValueStr = if (watchedValues == 0) "" else s" and $watchedValues other values"
+    val watchedValueStr =
+      if (watchedValueCount == 0) "" else s" and $watchedValueCount other values"
 
-    streams.err.println(
+    streams.err.println {
+      val viaFsNotify = if (useNotify) " (via fsnotify)" else ""
       colors.info(
-        s"Watching for changes to ${watchedPaths.size} paths$watchedValueStr... (Enter to re-run, Ctrl-C to exit)"
+        s"Watching for changes to ${watchedPathsSeq.size} paths$viaFsNotify$watchedValueStr... (Enter to re-run, Ctrl-C to exit)"
       ).toString
-    )
+    }
+
+    def doWatch(notifiablesChanged: () => Boolean) = {
+      val enterKeyPressed = statWatchWait(watchedPollables, stdin, notifiablesChanged)
+      setIdle(false)
+      enterKeyPressed
+    }
 
-    val enterKeyPressed = statWatchWait(watched, stdin)
-    setIdle(false)
-    enterKeyPressed
+    if (useNotify) Using.resource(os.write.outputStream(serverDir / "fsNotifyWatchLog")) { watchLog =>
+      def writeToWatchLog(s: String): Unit = {
+        watchLog.write(s.getBytes(java.nio.charset.StandardCharsets.UTF_8))
+        watchLog.write('\n')
+      }
+
+      @volatile var pathChangesDetected = false
+
+      // oslib watch only works with folders, so we have to watch the parent folders instead
+
+      writeToWatchLog(s"[watched-paths:unfiltered] ${watchedPathsSet.toSeq.sorted.mkString("\n")}")
+
+      val workspaceRoot = mill.api.WorkspaceRoot.workspaceRoot
+
+      /** Paths that are descendants of [[workspaceRoot]]. */
+      val pathsUnderWorkspaceRoot = watchedPathsSet.filter { path =>
+        val isUnderWorkspaceRoot = path.startsWith(workspaceRoot)
+        if (!isUnderWorkspaceRoot) {
+          streams.err.println(colors.error(
+            s"Watched path $path is outside workspace root $workspaceRoot, this is unsupported."
+          ).toString())
+        }
+
+        isUnderWorkspaceRoot
+      }
+
+      // If I have 'root/a/b/c'
+      //
+      // Then I want to watch:
+      //   root/a/b/c
+      //   root/a/b
+      //   root/a
+      //   root
+      val filterPaths = pathsUnderWorkspaceRoot.flatMap { path =>
+        path.relativeTo(workspaceRoot).segments.inits.map(segments => workspaceRoot / segments)
+      }
+      writeToWatchLog(s"[watched-paths:filtered] ${filterPaths.toSeq.sorted.mkString("\n")}")
+
+      Using.resource(os.watch.watch(
```

Comment:
This whole block of code should probably be moved into a separate helper method, maybe `doNotifyWatch`, in contrast to `doPollingWatch`